lib/SPIRV/SPIRVReader.cpp

```diff
@@ -4364,6 +4364,44 @@ bool llvm::getSpecConstInfo(std::istream &IS,
   return !IS.bad();
 }
 
+bool llvm::readSPIRVHeader(std::istream &IS, SPIRVHeaderData &Result,
+                           std::string &ErrMsg) {
+  std::unique_ptr<SPIRVModule> BM(SPIRVModule::createSPIRVModule());
+  BM->setAutoAddExtensions(false);
+  SPIRVDecoder D(IS, *BM);
+  SPIRVWord Magic;
+  SPIRVWord Buffer;
+  bool Success;
+  D >> Magic;
+  if (!BM->getErrorLog().checkError(Magic == MagicNumber, SPIRVEC_InvalidModule,
+                                    "invalid magic number")) {
+    ErrMsg = "invalid magic number";
+    return false;
+  }
+
+  D >> Buffer;
+  Result.Version = static_cast<VersionNumber>(Buffer);
+  // Skip the rest of the header
+  D.ignore(3);
+  while (D.getWordCountAndOpCode()) {
+    Success = false;
+    if (D.OpCode == OpCapability) {
+      Success = D.getNextSPIRVWord(Buffer);
+      if (Success)
+        Result.Capabilities.push_back(Buffer);
+    } else if (D.OpCode == OpMemoryModel) {
```

Comment:
If there is no MemoryModel instruction in a SPIR-V file then we will read a whole file. May there be some instruction after that there is no sense in searching MemoryModel and we can exit the loop earlier?

Reply:
I think if we encounter any instruction which is not one of `OpCapability`, `OpExtension`, `OpExtInstImport`, `OpMemoryModel` we can exit.